lib/ts/recipe/session/sessionAuth.tsx

```diff
@@ -30,142 +32,127 @@ type PropsWithoutAuth = {
 
-        return setContext({
-            doesSessionExist: sessionExists,
-            jwtPayload,
-            userId,
-        });
+            if (sessionExists === false) {
+                return {
+                    doesSessionExist: false,
+                    jwtPayload: {},
+                    userId: "",
+                };
+            }
+
+            return {
+                doesSessionExist: true,
+                jwtPayload: await getJWTPayloadSecurely(),
+                userId: await getUserId(),
+            };
+        };
+
+        async function setInitialContextAndMaybeRedirect() {
+            // if this component is unmounting, or the context has already
+            // been set, then we don't need to proceed...
+            if (cancelUseEffect || context !== undefined) {
+                return;
+            }
+
+            const toSetContext = await buildContext();
+
+            if (!toSetContext.doesSessionExist && props.requireAuth === true) {
+                redirectToLogin();
+            } else {
+                setContext(toSetContext);
+            }
+        }
+
+        setInitialContextAndMaybeRedirect();
+        return () => {
+            cancelUseEffect = true;
+        };
     }, []);
 
-    const onHandleEvent = useCallback(
-        async (event: RecipeEvent) => {
+    // subscribe to events on mount
+    useEffect(() => {
+        function onHandleEvent(event: RecipeEventWithSessionContext) {
             switch (event.action) {
                 case "SESSION_CREATED":
-                    if (!hasParentProvider(parentSessionContext)) {
-                        setContext({
-                            doesSessionExist: true,
-                            userId: await getUserId(),
-                            jwtPayload: await getJWTPayloadSecurely(),
-                        });
-                    }
+                    setContext(event.sessionContext);
                     return;
                 case "REFRESH_SESSION":
-                    if (!hasParentProvider(parentSessionContext)) {
-                        setContext({
-                            doesSessionExist: true,
-                            userId: context === undefined ? "" : context.userId,
-                            jwtPayload: await getJWTPayloadSecurely(),
-                        });
-                    }
+                    setContext(event.sessionContext);
                     return;
                 case "SIGN_OUT":
-                    if (!hasParentProvider(parentSessionContext)) {
-                        setContext({
-                            doesSessionExist: false,
-                            userId: "",
-                            jwtPayload: {},
-                        });
+                    if (props.requireAuth !== false) {
+                        setContext(event.sessionContext);
                     }
                     return;
                 case "UNAUTHORISED":
```

Comment:
This event can be fired multiple times, very fast - if there are many requests that are fired together and all yield `UNAUTHORISED`. This means that the `onSessionExpired` popup will be called many times. We want to avoid that by somehow keeping track of it has been called already or not.

I'm not sure what the best way to keep track of this is.. maybe a state, but then states update asynchronously. If this was a class component, we could have used an instance variable for this, and it would work fine.. not sure what the equivalent here would be.

Reply:
If it's designed to set some flag on parent component to display popup, then it should be idempotent, so multiple calls setting that flag to true shouldn't make a difference. I'd ignore that.

Reply:
We may want to mention it in the detailed docs though.

Reply:
Yes. Makes sense when `onSessionExpired` is called. However, if it's `redirectToLogin` being called, then that will be called multiple times. So maybe we want to set some state in the SessionAuth component and then on that state change, we want to redirect.. This will make calling `redirectToLogin` idempotent

Reply:
I see we have two options:
- either we instruct the user that this should be idempotent as well (set a flag to true that displays a redirection component)
- we make sure it’s called once by setting the state (which again increases the complexity of the component)

Let me know what you think

Reply:
From a user's point of view, making `redirectToLogin` idempotent is complicated.. so i prefer the second point.

Reply:
If we are doing this for `redirectToLogin`, we can apply the exact same logic for `onSessionExpired` as well.

Reply:
One thing to be careful about here is that if `requireAuth=false`, then `onSessionExpired` should be called just once whilst the session doesn't exist. But if the session then exists again, and then unauthorized again, then `onSessionExpired` should be called again.